staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -159,6 +160,47 @@ func (i *indexedWatchers) terminateAll(objectType reflect.Type, done func(*cache
 	}
 }
 
+// As we don't need a high precision here, we keep all watchers timeout within a
+// second in a bucket, and pop up them once at the timeout. To be more specific,
+// if you set fire time at X, you can get the bookmark within (X-1,X+1) period.
+type timeBucketWatchers struct {
+	watchersQueue map[int64][]*cacheWatcher
+	startBucketID int64
+	clock         clock.Clock
+}
+
+func newTimeBucketWatchers(clock clock.Clock) *timeBucketWatchers {
+	return &timeBucketWatchers{
+		watchersQueue: make(map[int64][]*cacheWatcher),
+		startBucketID: clock.Now().Unix(),
+		clock:         clock,
+	}
+}
+
+// Adding a watcher to the bucket, if the deadline is before the start, it will be
+// added to the first one.
+func (t *timeBucketWatchers) addWatcher(w *cacheWatcher) int64 {
+	bucketID := w.nextBookmarkTime(t.clock.Now()).Unix()
+	if bucketID < t.startBucketID {
+		bucketID = t.startBucketID
+	}
+	watchers, _ := t.watchersQueue[bucketID]
+	t.watchersQueue[bucketID] = append(watchers, w)
+	return bucketID
+}
+
+func (t *timeBucketWatchers) popExpiredWatchers() [][]*cacheWatcher {
+	currentBucketID := t.clock.Now().Unix()
+	var expiredWatchers [][]*cacheWatcher
```

Comment:
This means that we're allocating this slice in every single call to this function. But technically, it almost all cases there should be one or two elements (we're processing watchers with end_time in the last second), so maybe that's not that big deal...

Reply:
yeah, reserve 2 capacity seems better.